src/server/ftpserver.rs

```diff
@@ -256,21 +319,162 @@ where
+            // - incoming tcp connections originating from the proxy
+            // - channel messages originating from PASV, to handle the passive listening port
+
+            tokio::select! {
+
+                Some(tcp_stream) = incoming.next() => {
+                    let mut tcp_stream = tcp_stream.unwrap();
+                    let socket_addr = tcp_stream.peer_addr();
+
+                    // TODO: spawn a task here, to prevent blocking the main process
+                    info!("Incoming proxy connection from {:?}", socket_addr);
+                    let connection = match get_peer_from_proxy_header(&mut tcp_stream).await {
+                        Ok(v) => v,
+                        Err(e) => {
+                            warn!("proxy protocol decode error: {:?}", e);
+                            continue;
+                        }
+                    };
+
+                    // Based on the proxy protocol header, and the configured control port number,
+                    // we differentiate between connections for the control channel,
+                    // and connections for the data channel.
+                    if connection.to_port == proxy_params.external_control_port {
+                        let socket_addr = SocketAddr::new(connection.from_ip, connection.from_port);
+                        info!("Incoming control channel connection from {:?}", socket_addr);
+
+                        let result = self.spawn_control_channel_loop(tcp_stream, Some(connection), Some(proxyloop_msg_tx.clone())).await;
+                        if result.is_err() {
+                            warn!("Could not spawn control channel loop for connection: {:?}", result.err().unwrap())
+                        }
+                    } else {
+                        // handle incoming data connections
+                        println!("{:?}, {}", self.passive_ports, connection.to_port);
+                        if !self.passive_ports.contains(&connection.to_port) {
+                            error!("Incoming proxy connection going to unconfigured port! This port is not configured as a passive listening port: port {} not in passive port range {:?}", connection.to_port, self.passive_ports);
+                            tcp_stream.shutdown(Shutdown::Both).unwrap();
+                            continue;
+                        }
+
+                        self.dispatch_data_connection(tcp_stream, connection).await;
+
+                    }
+                },
+                Some(msg) = proxyloop_msg_rx.next() => {
+                    match msg {
+                        ProxyLoopMsg::AssignDataPortCommand (session_arc) => {
+                            self.select_and_register_passive_port(session_arc).await;
+                        },
+                    }
+                },
+            };
+        }
+    }
+
+    // this function finds (by hashing <srcip>.<dstport>) the session
```

Comment:
strictly speaking its not hashing :-) but rather creating a key. The hashtable will create a has from the key.